File: NEWCODER/JZ14.cpp
/**
 * 
 * 题目：
 * 输入一个链表，输出该链表中倒数第k个结点。
 * 示例：
 * 输入：
 * 1,{1,2,3,4,5}
 * 输出：
 * {5}
 * 
 * 要求：
 * 时间限制：C/C++ 1秒，其他语言2秒
 * 空间限制：C/C++ 64M，其他语言128M 
 */

#include<iostream>

using namespace std;

struct ListNode {
	int val;
	struct ListNode *next;
	ListNode(int x) :
			val(x), next(NULL) {
	}
};

//时间占用3ms | 内存占用376ms
ListNode* FindKthToTail(ListNode* pListHead, unsigned int k) {
    int len = 0;
    ListNode* p = pListHead;
    while(p){
            len++;
        p = p->next;
    }
    p = pListHead;
    int i = 0;
    while(p){
        i++;
        if((len-i+1)==(int)k)
            return p;
        p = p->next;
    }
    return nullptr;
}

//时间占用3ms | 内存占用476ms
ListNode* FindKthToTail_(ListNode* pListHead, unsigned int k) {
    if(!pListHead || k<=0) return nullptr;
    int len = 0;
    ListNode* p = pListHead;
    while(p){
        len++;
        p = p->next;
    }
    if(len < k) return nullptr;
    len -= k;
    while(len--){
        pListHead = pListHead->next;
    }
    return pListHead;
}

//快慢指针
ListNode* FindKthToTail__(ListNode* pListHead, unsigned int k) {
    if(!pListHead || k<=0) return nullptr;
    auto slow = pListHead, fast = pListHead;
    while (k--) {
        if(fast)
            fast = fast->next;
        else
            return nullptr;
    }
    while(fast) {
        slow = slow->next;
        fast = fast->next;
    }
    return slow;
}

int main(){
    ListNode* pListHead = new ListNode(1);
    ListNode* p = pListHead;
    p->next = new ListNode(2);
    p = p->next;
    p->next = new ListNode(3);
    p = p->next;
    p->next = new ListNode(4);
    p = p->next;
    p->next = new ListNode(5);
    cout << FindKthToTail(pListHead,2)->val << endl;
    cout << FindKthToTail_(pListHead,3)->val << endl;
    system("pause");
    return 1;
}

File: NEWCODER/JZOffer/Python/JZ14.py
# /**
#  *
#  * 题目：
#  * 输入一个链表，输出该链表中倒数第k个结点。
#  * 示例：
#  * 输入：
#  * 1,{1,2,3,4,5}
#  * 输出：
#  * {5}
#  *
#  * 要求：
#  * 时间限制：C/C++ 1秒，其他语言2秒
#  * 空间限制：C/C++ 64M，其他语言128M
#  */

class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


def FindKthToTail(head, k):
    """
    遍历全部链表，并保存
    TimeComplex o(n), SpaceComplex o(n)
    :param head:
    :param k:
    :return:
    """
    # write code here
    ans = []
    if head == None:
        return None
    else:
        while (head.next != None):
            ans.append(head)
            head = head.next
        ans.append(head)
    if k > len(ans) or k < 1:
        return None
    return ans[-k]


def FindKthToTail_(head, k):
    """
    双指针遍历
    TimeComplex o(n), SpaceComplex o(1)
    :param head:
    :param k:
    :return:
    """
    # write code here
    slow, fast = head, head
    for i in range(k):
        if not fast:
            return None
        fast = fast.next
    while fast:
        slow = slow.next
        fast = fast.next
    return slow


if __name__ == '__main__':
    head = ListNode(1)
    head.next = ListNode(2)
    head.next.next = ListNode(3)
    head.next.next.next = ListNode(4)
    head.next.next.next.next = ListNode(5)
    head.next.next.next.next.next = None

    # print(FindKthToTail([1, 2, 3, 4, 5, 6, 7, 8]))
    # print(FindKthToTail(head,2).val)
    print(FindKthToTail_(head, 2).val)